File: src/config/loader.ts
import fs from "node:fs";
import { getConfigPath } from "./paths";
import { type AppConfig, AppConfigSchema } from "./schema";

export function loadConfig(customPath?: string): AppConfig {
	const cfgPath = customPath || getConfigPath();
	if (!fs.existsSync(cfgPath)) {
		return AppConfigSchema.parse({});
	}
	try {
		const raw = fs.readFileSync(cfgPath, "utf-8");
		return AppConfigSchema.parse(JSON.parse(raw));
	} catch (e) {
		console.warn(`Failed to load config: ${(e as Error).message}`);
		return AppConfigSchema.parse({});
	}
}

export function saveConfig(config: AppConfig, customPath?: string): void {
	const cfgPath = customPath || getConfigPath();
	fs.writeFileSync(cfgPath, JSON.stringify(config, null, 2), "utf-8");
}

File: miniclaw/config/loader.py
"""Configuration loader for Miniclaw."""

import json
from pathlib import Path
from loguru import logger
from .schema import AppConfig
from .paths import get_config_path


def load_config(path: Path | None = None) -> AppConfig:
    cfg_path = path or get_config_path()
    if not cfg_path.exists():
        return AppConfig()
    try:
        with open(cfg_path, "r", encoding="utf-8") as f:
            data = json.load(f)
        return AppConfig.model_validate(data)
    except Exception as e:
        logger.warning(f"Failed to load config: {e}")
        return AppConfig()


def save_config(config: AppConfig, path: Path | None = None) -> None:
    cfg_path = path or get_config_path()
    cfg_path.parent.mkdir(parents=True, exist_ok=True)
    data = config.model_dump(mode="json", exclude_none=True)
    with open(cfg_path, "w", encoding="utf-8") as f:
        json.dump(data, f, indent=2)
